exporter/sumologicexporter/exporter.go

```diff
@@ -16,44 +16,273 @@ package sumologicexporter
+		skip := !filterOut
+		for j := 0; j < len(se.metadataRegexes); j++ {
+			if se.metadataRegexes[j].MatchString(k) {
+				skip = false
+				if filterOut {
+					return
+				}
+			}
+		}
+
+		if skip {
+			return
+		}
+
+		returnValue[k] = v.StringVal()
+	})
+
+	return returnValue
 }
 
-func (se *sumologicexporter) ConsumeLogs(_ context.Context, _ pdata.Logs) error {
+// GetMetadata builds string which represents metadata in alphabetical order
+func (se *sumologicexporter) GetMetadata(attributes pdata.AttributeMap) string {
+	attrs := se.filterMetadata(attributes, false)
+	keys := make([]string, 0, len(attrs))
+	buf := strings.Builder{}
+
+	for k := range attrs {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+
+	for _, k := range keys {
+		if buf.Len() > 0 {
+			buf.WriteString(", ")
+		}
+		buf.WriteString(fmt.Sprintf("%s=%s", k, attrs[k]))
+	}
+
+	return buf.String()
+}
+
+// This function tries to send data and modify pass values
+func (se *sumologicexporter) sendAndPushErrors(buffer *[]pdata.LogRecord, fields string, droppedTimeSeries *int, errs *[]error) {
+	err := se.sendLogs(*buffer, fields)
+	if err != nil {
+		*droppedTimeSeries += len(*buffer)
+		*errs = append(*errs, err)
+	}
+	*buffer = (*buffer)[:0]
+}
+
+// pushLogsData groups data with common metadata uses Send to send data to sumologic
+func (se *sumologicexporter) pushLogsData(ctx context.Context, ld pdata.Logs) (droppedTimeSeries int, err error) {
+	maxBufferSize := 100
+	buffer := make([]pdata.LogRecord, 0, maxBufferSize)
+	previousMetadata := ""
+	currentMetadata := ""
+	var errs []error
+
+	// Iterate over ResourceLogs
+	for i := 0; i < ld.ResourceLogs().Len(); i++ {
+		resource := ld.ResourceLogs().At(i)
+
+		// iterate over InstrumentationLibraryLogs
+		for j := 0; j < resource.InstrumentationLibraryLogs().Len(); j++ {
+			library := resource.InstrumentationLibraryLogs().At(j)
+
+			// iterate over Logs
+			for k := 0; k < library.Logs().Len(); k++ {
+				log := library.Logs().At(k)
+				currentMetadata = se.GetMetadata(log.Attributes())
+
+				// If metadate differs from currently buffered, flush the buffer
+				if currentMetadata != previousMetadata && previousMetadata != "" {
+					se.sendAndPushErrors(&buffer, previousMetadata, &droppedTimeSeries, &errs)
+				}
+
+				// assign metadata
+				previousMetadata = currentMetadata
+
+				// add log to the buffer
+				buffer = append(buffer, log)
+
+				// Flush buffer to avoid overlow
+				if len(buffer) == maxBufferSize {
+					se.sendAndPushErrors(&buffer, previousMetadata, &droppedTimeSeries, &errs)
+				}
+			}
+		}
+	}
+
+	// Flush pending logs
+	se.sendAndPushErrors(&buffer, previousMetadata, &droppedTimeSeries, &errs)
+
+	return droppedTimeSeries, componenterror.CombineErrors(errs)
+}
+
+func (se *sumologicexporter) sendLogs(buffer []pdata.LogRecord, fields string) error {
+	body := strings.Builder{}
+	var errs []error
+
+	if se.config.LogFormat == TextFormat {
```

Comment:
This whole `if` is a lot to digest (and a many indentation levels to unfold in one's head :) ).
Let's extract the inner parts of this `if` into separate functions.

Reply:
done